packages/taquito/src/provider.ts

```diff
@@ -0,0 +1,202 @@
+import { Context } from './context';
+import { ForgedBytes, ParamsWithKind, RPCOperation, isOpRequireReveal } from './operations/types';
+import {
+  InvalidEstimateValueError,
+  TezosOperationError,
+  TezosPreapplyFailureError,
+  flattenErrors,
+} from './operations/operation-errors';
+import {
+  createOriginationOperation,
+  createRegisterGlobalConstantOperation,
+  createSetDelegateOperation,
+  createTransferOperation,
+  createTransferTicketOperation,
+  createIncreasePaidStorageOperation,
+  createSmartRollupAddMessagesOperation,
+  createSmartRollupOriginateOperation,
+} from './contract/prepare';
+import { OpKind } from '@taquito/rpc';
+import { InvalidOperationKindError } from '@taquito/utils';
+import { PreparedOperation } from './prepare';
+import { Estimate } from './estimate';
+
+export abstract class Provider {
+  get rpc(): RpcClientInterface {
+    return this.context.rpc;
+  }
+
+  get signer() {
+    return this.context.signer;
+  }
+
+  constructor(protected context: Context) {}
+
+  protected async forge({ opOb: { branch, contents, protocol }, counter }: PreparedOperation) {
+    const forgedBytes = await this.context.forger.forge({ branch, contents });
+    return {
+      opbytes: forgedBytes,
+      opOb: {
+        branch,
+        contents,
+        protocol,
+      },
+      counter,
+    };
+  }
+
+  protected async estimate<T extends { fee?: number; gasLimit?: number; storageLimit?: number }>(
+    { fee, gasLimit, storageLimit, ...rest }: T,
+    estimator: (param: T) => Promise<Estimate>
+  ) {
+    let calculatedFee = fee;
+    let calculatedGas = gasLimit;
+    let calculatedStorage = storageLimit;
+
+    if (calculatedFee && calculatedFee % 1 !== 0) {
+      throw new InvalidEstimateValueError(`Fee value must not be a decimal: ${calculatedFee}`);
+    }
+    if (calculatedGas && calculatedGas % 1 !== 0) {
+      throw new InvalidEstimateValueError(
+        `Gas Limit value must not be a decimal: ${calculatedGas}`
+      );
+    }
+    if (calculatedStorage && calculatedStorage % 1 !== 0) {
+      throw new InvalidEstimateValueError(
+        `Storage Limit value must not be a decimal: ${calculatedStorage}`
+      );
+    }
+
+    if (fee === undefined || gasLimit === undefined || storageLimit === undefined) {
+      const estimation = await estimator({ fee, gasLimit, storageLimit, ...(rest as any) });
+
+      if (calculatedFee === undefined) {
+        calculatedFee = estimation.suggestedFeeMutez;
+      }
+
+      if (calculatedGas === undefined) {
+        calculatedGas = estimation.gasLimit;
+      }
+
+      if (calculatedStorage === undefined) {
+        calculatedStorage = estimation.storageLimit;
+      }
+    }
+
+    return {
+      fee: calculatedFee,
+      gasLimit: calculatedGas,
+      storageLimit: calculatedStorage,
+    };
+  }
+
+  async getRPCOp(param: ParamsWithKind) {
+    switch (param.kind) {
```

Comment:
This switch statement does not cover `TX_ROLLUP_ORIGINATION`, `TX_ROLLUP_SUBMIT_BATCH`, and `UPDATE_CONSENSUS_KEY`.

Reply:
tx operations are deprecated, whereas I don't think UPDATE_CONSENSUS_KEY is batchable in this case

Reply:
same missing reveal and updateConsensusKey (understand if reveal is not needed)

Reply:
understand to not support batch updateConsensusKey in taquito from previous thread